wire-runtime/src/main/java/com/squareup/wire/ProtoWriter.java

```diff
@@ -107,45 +94,24 @@ static int varint64Size(long value) {
     return 10;
   }
 
-  /**
-   * Encode a ZigZag-encoded 32-bit value. ZigZag encodes signed integers into values that can be
-   * efficiently encoded with varint. (Otherwise, negative values must be sign-extended to 64 bits
-   * to be varint encoded, thus always taking 10 bytes on the wire.)
-   *
-   * @param n A signed 32-bit integer.
-   * @return An unsigned 32-bit integer, stored in a signed int because Java has no explicit
-   * unsigned support.
-   */
-  static int zigZag32(int n) {
-    // Note:  the right-shift must be arithmetic
-    return (n << 1) ^ (n >> 31);
-  }
-
-  /**
-   * Encode a ZigZag-encoded 64-bit value. ZigZag encodes signed integers into values that can be
-   * efficiently encoded with varint. (Otherwise, negative values must be sign-extended to 64 bits
-   * to be varint encoded, thus always taking 10 bytes on the wire.)
-   *
-   * @param n A signed 64-bit integer.
-   * @return An unsigned 64-bit integer, stored in a signed int because Java has no explicit
-   * unsigned support.
-   */
-  static long zigZag64(long n) {
-    // Note:  the right-shift must be arithmetic
-    return (n << 1) ^ (n >> 63);
-  }
-
   private final BufferedSink sink;
 
   ProtoWriter(BufferedSink sink) {
     this.sink = sink;
   }
 
-  void writeRawByte(int value) throws IOException {
+  public <T> void write(TypeAdapter<T> adapter, T value) throws IOException {
+    if (adapter.type == WireType.LENGTH_DELIMITED) {
+      writeVarint32(adapter.serializedSize(value));
```

Comment:
There's a problem with that which is the root message, a length-delimited object, should NOT prefix itself with its length. This could be solved by adding state to `ProtoWriter` to ignore the first call to `writeVarint32` or something... but that didn't seem great either.